### DISCUSSION TITLE
Occasional random Query Error Code 1317?

### DISCUSSION THREAD

**Discussion by MaxBittker** [planetscale/discussion#219] [Bug Reports] [answered]
I'm occasionally seeing this error, even for very simple queries that usually have an execution time of <5ms. 
Any idea what I can do to prevent or understand this? Can i provide more info? :) 

```
[GET] /api/getCreation/770
01:07:39:24
2022-06-15T05:07:39.281Z	955606b3-c4b0-4c86-b8bd-b445a01f4376	ERROR	PrismaClientUnknownRequestError: Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1317, message: "vttablet: rpc error: code = Aborted desc = transaction 1655216299250051223: ended at 2022-06-15 05:03:32.216 UTC (exceeded timeout: 20s) (CallerID: planetscale-admin)", state: "70100" })) })
```

**Comment by jgreet** [comment, marked as answer]
Hi @MaxBittker, the 20s timeout limits open transaction times, so even if the individual queries executed within the transaction are very fast, you can still hit it if the app goes more than 20 seconds without rolling back or committing the transaction.

For example, it's easy to reproduce the error message from the timeout by waiting 20+ seconds:

```mysql
beam/|⚠ main ⚠|> start transaction; do sleep (21); select now();
ERROR 1317 (70100): vttablet: rpc error: code = Aborted desc = transaction 1655222528817681011: ended at 2022-06-17 19:19:52.923 UTC (unlocked closed connection) (CallerID: planetscale-admin)
```

So the first things I'd look for are long running transactions, or code that could open but fail to close a transaction. Enabling [query logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging) in Prisma might help you figure out where that's happening.

**Comment by underhodler** [comment]
Hi, I am doing some test running the identical query with <5ms runtime, I am not running any app just trying to run the query itself and most of the time it succeed but sometimes it randomly throws error: `primary: vttablet: rpc error: code = Aborted desc = transaction 1681784895267363563: ended at 2023-04-20 03:32:09.508 UTC (exceeded timeout: 20s) (CallerID: pfn1w1lv2fh77ff2elok)')`.
Do you know what could be the reason?

**Reply by fooforge** [reply]
Most likely, it's for the same reason as mentioned in https://github.com/planetscale/discussion/discussions/408#discussioncomment-5674020. We have some deliberately set [system limits](https://planetscale.com/docs/reference/planetscale-system-limits) including a 20 seconds timeout on transactions. I assume, you're using an ORM that wraps the query inside a transaction?

**Reply by underhodler** [reply]
Thanks for the pointer. I am just using database connector library i.e. MySQLdb with python. To give more context, I was running the exact same query and only sometimes it failed with this timeout. 

There is another case where I tried to insert 1 row via python and it showed inserted successfully (I was able to query in the console and see the new record appeared in the table) but after ~20 seconds when I queried again the record is gone and I get this timeout thing again when I queried the table via python. 

It's weird how a record showed up in the table and disappeared soon after. Any help would be super appreciated!

**Reply by mattwalshdev** [reply]
Hi underhodler,
This page came up in a google search and I've got the same issue, not using Prisma, just PlanetScale. Just worked it out, it's a bug with the MySQLdb python package and Vitess. I kept having the same issue, where I have a 20s+ running python with an open db connection, BUT I am making and committing sub-1second transactions throughout. The solution at this point is to use the pymysql library instead as it doesn't run into this issue.

Edit - not a bug but a missing argument, if you use MySQLdb python library and keep a PlanetScale connection open for 20+seconds you need to add `autocommit=True` to you connection arguments, or make sure you commit after all operations, **including SELECT** see https://github.com/planetscale/discussion/discussions/557
